### PR TITLE
Add room storage for Django Channels Consumers

### PR DESCRIPTION
**Pull request by cacosandon**
This PR introduces a shared storage for consumers within Django Channels.

Initially, each YjsConsumer had its unique YDoc, updating only from its client. This approach caused issues, such as consumers having different YDoc versions that made difficult to update to a persistent storage.

Now, by inheriting from BaseYRoomStorage, a unified storage can be utilized across all consumers and server-wide (e.g., in a Django Celery Job).

The storage's function, detailed in the documentation, is to manage document changes—fetching, updating, and saving them to a single source. An implementation example features Redis for ephemeral and Postgres for permanent storage.

Each user can define their own storage. Maybe using Django Cache, custom Redis or no storage at all (and just use it to broadcast updates from their connected clients). We just define the interface.

### CODE REVIEW

**Comment by cacosandon** [comment]
Running `mypy` to reproduce the "Check types" phase of the tests generates a lot of errors, the majority of them related to missing stubs.

How are you solving them?

**Comment by davidbrochart** [comment]
Thanks @cacosandon for the PR.
At first sight, I'm wondering if you're not reimplemention the [ystores](https://github.com/jupyter-server/pycrdt-websocket/blob/main/pycrdt_websocket/ystore.py)?

**Comment by cacosandon** [comment]
> Thanks @cacosandon for the PR. At first sight, I'm wondering if you're not reimplemention the [ystores](https://github.com/jupyter-server/pycrdt-websocket/blob/main/pycrdt_websocket/ystore.py)?

Hey @davidbrochart!

The core distinction between `YStore` (and `YRoom`) and Django Channels lies in their operational contexts. As I see, `YStore` is designed for a single-server environment, orchestrating multiple rooms with explicit start and stop controls. This centralized model enables unified management of YDoc instances across rooms.

Conversely, Django Channels employs a decentralized model, where each consumer operates in an independent execution thread. This necessitates a reevaluation of state management for shared states, facilitating concurrent document updates and synchronization. Our approach provides an "ephemeral" storage solution, keeping everyone in sync without manual management of each document's store lifecycle.

In summary, both systems aim to synchronize and persist document states, but the decentralized nature of Django Channels necessitates a tailored solution for synced independent instances. I mean, we can adapt the current setup to use `YRoom` and `YStore`, but we'll likely use only the functions for getting snapshots, saving, and updating, etc., not the lifecycle methods for which both classes are intended. Also, because there is no centralized document, we need this "ephemeral" storage, which is basically holds this shared YDoc (that could be a global variable, a file, or Redis as we show as an example).

**Comment by cacosandon** [comment]
> Thanks @cacosandon for the PR. At first sight, I'm wondering if you're not reimplemention the [ystores](https://github.com/jupyter-server/pycrdt-websocket/blob/main/pycrdt_websocket/ystore.py)?

Hey @davidbrochart! Sorry for the delay.

The core distinction between `YStore` (and `YRoom`) and the Django Channels implementation lies in their operational contexts. As I see, `YStore` is designed for a single-server environment, orchestrating multiple rooms with explicit start and stop controls. This centralized model enables unified management of YDoc instances across rooms.

On the other side, Django Channels employs a decentralized model, where each consumer operates in an independent execution thread. Our approach provides an "ephemeral" storage solution, keeping everyone in sync without manual management of each document's store lifecycle.

In summary, both systems aim to synchronize and persist document states, but the decentralized nature of Django Channels needs a YDoc synced outside the server. I mean, we can adapt the current setup to use `YRoom` and `YStore`, but we'll likely use only the functions for getting snapshots, saving, updating, etc., not the lifecycle methods for which both classes are intended. Also, because there is no centralized document, we need this "ephemeral" storage, which basically holds this shared YDoc (that could be a global variable, a file, or Redis as we show as an example).

**Comment by cacosandon** [comment]
Hey @davidbrochart, have you taken a look at this? 🙏

File: pycrdt_websocket/django_channels/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from .yjs_consumer import YjsConsumer as YjsConsumer
+from .yroom_storage import BaseYRoomStorage as BaseYRoomStorage
+from .yroom_storage import RedisYRoomStorage as RedisYRoomStorage
```
Comment (davidbrochart) [inline pycrdt_websocket/django_channels/__init__.py:3]: Does this mean that redis is a required dependency when using Django channels?

File: pycrdt_websocket/django_channels/yroom_storage.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
+import time
+from typing import Optional
+
+import redis.asyncio as redis
+from pycrdt import Doc
+
+
+class BaseYRoomStorage:
```
Comment (davidbrochart) [inline pycrdt_websocket/django_channels/yroom_storage.py:8]: Maybe this should be an abstract base class, with a dummy implementation that does nothing?

File: pycrdt_websocket/django_channels/yroom_storage.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                current_snapshot,
+            )
+    ```
+    """
+
+    def __init__(self, room_name: str) -> None:
+        self.room_name = room_name
+
+        self.last_saved_at = time.time()
+        self.save_throttle_interval = 5
```
Comment (davidbrochart) [inline pycrdt_websocket/django_channels/yroom_storage.py:62]: I don't think this should be part of the base class. Also it should be configurable.

**Review by davidbrochart — CHANGES_REQUESTED** [review]
Sorry for the delay @cacosandon.
I'm not sure about these changes, I still think that YRoomStorage is similar to YStore.

File: pycrdt_websocket/django_channels/__init__.py (hunk 1 above)
Comment (cacosandon) [reply pycrdt_websocket/django_channels/__init__.py]: Oh, right. I've removed `RedisYRoomStorage` from this init file and separate them in files so when you import the base class you are not requiring redis. Maybe we should do the same with `sqlite-anyio` on YStore, which is a dependency for the whole project.

File: pycrdt_websocket/django_channels/yroom_storage.py (hunk 2 above)
Comment (cacosandon) [reply pycrdt_websocket/django_channels/yroom_storage.py]: Agree!

File: pycrdt_websocket/django_channels/yroom_storage.py (hunk 3 above)
Comment (cacosandon) [reply pycrdt_websocket/django_channels/yroom_storage.py]: Agree

**Comment by cacosandon** [comment]
> Sorry for the delay @cacosandon. I'm not sure about these changes, I still think that YRoomStorage is similar to YStore.

Thank you, @davidbrochart, for the comments!

I think the primary distinction lies in our need for a shared `YDoc` that isn't directly within an instantiated class, as the repository's main server does. This shared doc is facilitated through RoomStorage.

While they are very similar—implementing the same methods for applying updates and writing—what they lack is an "ephemeral" storage to manage each of the updates from clients (or consumers in Django Channels).

Additionally, we don't require the property `started` or `__aenter__`, `__aexit__`, `start`, and `stop` methods (because Django Channels it's a decentralized model), but we do need others to synchronize the updates before writing to the database (if necessary).

Thus, while we could use `YStore`, duplicating it in this case to manage each separately seems preferable. But I am still open to that possibility.

Let me know what you think, and thanks again :)

**Comment by davidbrochart** [comment]
> While they are very similar—implementing the same methods for applying updates and writing—what they lack is an "ephemeral" storage to manage each of the updates from clients (or consumers in Django Channels).

There could be a new in-memory implementation of a YStore.

> Additionally, we don't require the property `started` or `__aenter__`, `__aexit__`, `start`, and `stop` methods (because Django Channels it's a decentralized model), but we do need others to synchronize the updates before writing to the database (if necessary).

Usually services have to do some kind of initialization and teardown, but if you don't need that then nothing prevents you from doing nothing in these methods.

> Thus, while we could use `YStore`, duplicating it in this case to manage each separately seems preferable.

But that's also a maintenance effort that I'd rather not do.
